language/move-core/types/src/unit_tests/value_test.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    account_address::AccountAddress,
+    ident_str,
+    identifier::Identifier,
+    language_storage::{StructTag, TypeTag},
+    value::{MoveStruct, MoveValue},
+};
+use serde_json::json;
+
+#[test]
+fn struct_deserialization() {
+    let struct_type = StructTag {
+        address: AccountAddress::ZERO,
+        name: ident_str!("MyStruct").to_owned(),
+        module: ident_str!("MyModule").to_owned(),
+        type_params: vec![],
+    };
+    let values = vec![MoveValue::U64(7), MoveValue::Bool(true)];
+    let fields = vec![ident_str!("f").to_owned(), ident_str!("g").to_owned()];
+    let field_values: Vec<(Identifier, MoveValue)> =
+        fields.into_iter().zip(values.clone()).collect();
+
+    // test each deserialization scheme
+    let runtime_value = MoveStruct::Runtime(values);
+    assert_eq!(
+        serde_json::to_value(&runtime_value).unwrap(),
+        json!([7, true])
+    );
+
+    let fielded_value = MoveStruct::WithFields(field_values.clone());
+    assert_eq!(
+        serde_json::to_value(&fielded_value).unwrap(),
+        json!({ "f": 7, "g": true })
+    );
+
+    let typed_value = MoveStruct::with_types(struct_type.clone(), field_values.clone());
+    assert_eq!(
+        serde_json::to_value(&typed_value).unwrap(),
+        json!({
+                "fields": { "f": 7, "g": true },
+                "type": "0x0::MyModule::MyStruct"
```

Comment:
I'm wondering if this type should be even further deconstructed possibly. e.g.,

```json
"type": {
    "address": "0x0",
    "module_name": "MyModule",
    "type_name": "MyStruct"
}
```

In the monomorphic case I don't think there is much of a difference, but in the generic case it may make it easier to distinguish/find multiple instantiations of the same generic type:

```json
[{
  "fields": { ... },
  "type": {
     "address": "0x0",
     "module_name": "MyModule",
     "type_name": "GenericStruct",
     "instantiation": [ "u8" ]
   }
},
{
  "fields": { ... },
  "type": {
     "address": "0x0",
     "module_name": "MyModule",
     "type_name": "GenericStruct",
     "instantiation": [ "u64" ]
   }
}]
```

Reply:
Very good question. I actually did it this way initially, but ended up changing my mind and going with the current approach. My reasoning:
- I think clients of the code will most commonly want to use the fully qualified type directly rather than breaking it into its components. E.g.: map the type to its Python or JS equivalent + create a Python or JS object for the value, index the struct value by its type, pipe it into a query like "show me other values of this type", copy the type and paste it into a block explorer. or just view the type in a human-readable format.
- This way ends up being quite a bit more human-readable, especially considering that every struct in a (potentially deep) nesting has an attached type. If you use "type as object representation", I think any nested struct quickly becomes quite difficult to grok.

With that said, I think we should probably support multiple ways of printing the types just as we support multiple ways of printing struct values. Some refactoring of value.rs is probably needed to support this, the current approach is already becoming a bit unwieldy...

Reply:
Good points, and agreed that this is becoming a bit unwieldy. Eventually I think it would be best to have a config where these types of things can be set ("packed" vs "unpacked" type representations in the output).